lighthouse-core/test/gather/gatherers/image-elements-test.js

```diff
@@ -0,0 +1,282 @@
+  return {
+    ...gatherer,
+    _getArtifact: jest.fn(makeParamsOptional(gatherer._getArtifact)),
+    collectExtraDetails: jest.fn(makeParamsOptional(gatherer.collectExtraDetails)),
+    indexNetworkRecords: jest.fn(makeParamsOptional(gatherer.indexNetworkRecords)),
+    fetchSourceRules: jest.fn(makeParamsOptional(gatherer.fetchSourceRules)),
+    fetchElementWithSizeInformation: jest.fn(
+      makeParamsOptional(gatherer.fetchElementWithSizeInformation)
+    ),
+  };
+}
+
+describe('.indexNetworkRecords', () => {
+  it('maps image urls to network records', () => {
+    const gatherer = mockImageElements();
+    const networkRecords = [
+      {
+        mimeType: 'image/png',
+        url: 'https://example.com/img.png',
+        finished: true,
+        statusCode: 200,
+      },
+      {
+        mimeType: 'application/octect-stream',
+        url: 'https://example.com/img.webp',
+        finished: true,
+        statusCode: 200,
+      },
+      {
+        mimeType: 'application/octect-stream',
+        url: 'https://example.com/img.avif',
+        finished: true,
+        statusCode: 200,
+      },
+    ];
+
+    const index = gatherer.indexNetworkRecords(networkRecords);
+
+    expect(index).toEqual({
+      'https://example.com/img.avif': {
+        finished: true,
+        mimeType: 'application/octect-stream',
+        statusCode: 200,
+        url: 'https://example.com/img.avif',
+      },
+      'https://example.com/img.png': {
+        finished: true,
+        mimeType: 'image/png',
+        statusCode: 200,
+        url: 'https://example.com/img.png',
+      },
+      'https://example.com/img.webp': {
+        finished: true,
+        mimeType: 'application/octect-stream',
+        statusCode: 200,
+        url: 'https://example.com/img.webp',
+      },
+    });
+  });
+
+  it('ignores bad status codes', () => {
+    const gatherer = mockImageElements();
+    const networkRecords = [
+      {
+        mimeType: 'image/png',
+        url: 'https://example.com/img.png',
+        finished: true,
+        statusCode: 200,
+      },
+      {
+        mimeType: 'application/octect-stream',
+        url: 'https://example.com/img.webp',
+        finished: false,
+      },
+      {
+        mimeType: 'application/octect-stream',
+        url: 'https://example.com/img.avif',
+        finished: true,
+        statusCode: 404,
+      },
+    ];
+
+    const index = gatherer.indexNetworkRecords(networkRecords);
+
+    expect(index).toEqual({
+      'https://example.com/img.png': {
+        finished: true,
+        mimeType: 'image/png',
+        statusCode: 200,
+        url: 'https://example.com/img.png',
+      },
+    });
+  });
+});
+
+describe('.collectExtraDetails', () => {
+  let gatherer = mockImageElements();
+
+  beforeEach(() => {
+    gatherer = mockImageElements();
+    gatherer.fetchSourceRules.mockImplementation();
+    gatherer.fetchElementWithSizeInformation.mockImplementation();
+  });
+
+  it('times out', () => {
```

Comment:
~~I'm not sure I understand what this one is supposed to be testing. isn't the only assertion here going to be true regardless of how timeout behavior is handled?~~

ah I gotcha :)
```suggestion
  it('respects the overall time budget for source rules', () => {
```